pythia/utils/text_utils.py

```diff
@@ -271,21 +261,93 @@ def search(self, t, data, scores):
 
         # TODO: Make the data update generic for any type of model
         # This is specific to BUTD model only.
-        data["texts"] = next_word_inds[incomplete_inds].unsqueeze(1)
-        h1 = data["state"]["td_hidden"][0][prev_word_inds[incomplete_inds]]
-        c1 = data["state"]["td_hidden"][1][prev_word_inds[incomplete_inds]]
-        h2 = data["state"]["lm_hidden"][0][prev_word_inds[incomplete_inds]]
-        c2 = data["state"]["lm_hidden"][1][prev_word_inds[incomplete_inds]]
-        data["state"] = {"td_hidden": (h1, c1), "lm_hidden": (h2, c2)}
+        data = update_data(data, prev_word_inds, next_word_inds, incomplete_inds)
 
         next_beam_length = len(prev_word_inds[incomplete_inds])
 
         return False, data, next_beam_length
 
-    def best_score(self):
+    def get_captions(self):
         if len(self.complete_seqs_scores) == 0:
             captions = torch.FloatTensor([0] * 5).unsqueeze(0)
         else:
             i = self.complete_seqs_scores.index(max(self.complete_seqs_scores))
             captions = torch.FloatTensor(self.complete_seqs[i]).unsqueeze(0)
         return captions
+
+class NucleusSampling:
+    def __init__(self, vocab, threshold=0.8):
+        self.vocab = vocab
+        self.vocab_size = vocab.get_size()
+        # Threshold for sum of probability
+        self.threshold = threshold
+        # Lists to store completed sequence
+        self.complete_seq = []
+
+    def init_batch(self, sample_list):
+        setattr(
```

Comment:
Any specific reason to use `setattr`?

Reply:
This pattern was reused from Beam Search.

Reply:
Hmm, but since `"seq"` is not a variable, directly assign it.